synapse/app/synctl.py

```diff
@@ -190,7 +216,22 @@ def main():
         if start_stop_synapse:
             stop(pidfile, "synapse.app.homeserver")
 
-        # TODO: Wait for synapse to actually shutdown before starting it again
+    # Wait for synapse to actually shutdown before starting it again
+    if action == "restart" and start_stop_synapse:
```

Comment:
you need to look more carefully at what `start_stop_synapse` is for. We may need to restart the workers even if `start_stop_synapse` is false.